### PR TITLE
Change how no-arg constructor is handled internally in Variant.

### LINKED ISSUE #6977: Expose built-in types default constructors
- Vector2
- Vector3
- Matrix32
- Matrix3
- Quaternion
- Color
- Transform
- ...

Primitive math types often have multiple constructors, but the default one is not listed. Also, some people read the docs first before writing code (especially with a scripting API), it can be misleading to hint there is no default constructor, leading them to not even try.

These constructors exist, they can be used. But if they are not found by doctool... then we have another problem, because not a single default constructor is ever documented, not just math primitives. Which leads me to think they are bound automatically, and unavailable to doctool.

Original request: https://github.com/godotengine/godot/pull/6752#issuecomment-257141732

### PR DESCRIPTION
**Pull request by reduz**
*Bugsquad edit:* Fully fixes #6977.

### CODE REVIEW

**Review by akien-mga — APPROVED** [review]
My own fix was less code :P
```diff
@@ -766,7 +771,10 @@ void Variant::get_constructor_list(Type p_type, List<MethodInfo> *r_list) {
        MethodInfo mi;
        mi.return_val.type = p_type;
        mi.name = get_type_name(p_type);
+       // Built-in constructor (e.g. `String()`).
+       r_list->push_back(mi);
 
+       // Constructors with arguments (e.g. `String(int)`).
        for (int i = 0; i < get_constructor_count(p_type); i++) {
                int ac = get_constructor_argument_count(p_type, i);
                mi.arguments.clear();
```

But yeah that looks good :)

File: core/variant/variant_construct.cpp [hunk 1]
```diff
@@ -601,76 +682,41 @@ void Variant::_register_variant_constructors() {
...
 
+	add_constructor<VariantConstructNoArgs<PackedInt32Array>>(sarray());
 	add_constructor<VariantConstructor<PackedInt32Array, PackedInt32Array>>(sarray("from"));
 	add_constructor<VariantConstructorFromArray<PackedInt32Array>>(sarray("from"));
 
+	add_constructor<VariantConstructNoArgs<PackedInt64Array>>(sarray());
 	add_constructor<VariantConstructor<PackedInt64Array, PackedInt64Array>>(sarray("from"));
 	add_constructor<VariantConstructorFromArray<PackedInt64Array>>(sarray("from"));
 
+	add_constructor<VariantConstructNoArgs<PackedInt32Array>>(sarray());
```
Comment (akien-mga) [inline core/variant/variant_construct.cpp:697]: This one is wrong, should be `Float32`.

Comment (akien-mga) [reply core/variant/variant_construct.cpp]: I force pushed an amend to fix it.

Comment (reduz) [reply core/variant/variant_construct.cpp]: oops, fixed

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**